### ISSUE TITLE
census only updating in 1 direction

### ISSUE THREAD

**Issue by bodymindarts** [habitat-sh/habitat#2016]
I have encountered a very strange bug that I am not able to pin down. 
When starting the dockerized habitat packages from [this repo](https://github.com/starkandwayne/habitat-plans/tree/census-bug) with [this docker-compose file](https://github.com/starkandwayne/habitat-plans/blob/census-bug/shield/test/docker-compose.yml) curling the http endpoints and inspecting the output to `/census` returns inconsistent results.

The pg endpoint contains information for both pg and shield the shield endpoint however only returns information about itself.

Postgres endpoint:
```json
$ curl localhost:9631/census | jq
{
  "censuses": {
    "postgresql.shield": {      
      "population": {
        "31efd0e3225b4e4999dee26ab5de3c9e": {
          "member_id": "31efd0e3225b4e4999dee26ab5de3c9e",
          "service": "postgresql",
          "group": "shield",
          "org": null,
          "cfg": {
            "port": 5432,
            "superuser_name": "admin",
            "superuser_password": "admin"
          },
          "sys": {
            "ip": "172.26.0.2",
            "hostname": "9aa90df39e26",
            "gossip_ip": "172.26.0.2",
            "gossip_port": "9638",
            "http_gateway_ip": "0.0.0.0",
            "http_gateway_port": "9631"
          },
          "pkg": {
            "origin": "starkandwayne",
            "name": "postgresql",
            "version": "9.6.1",
            "release": "20170329081713"
          },
          "leader": null,
          "follower": null,
          "update_leader": null,
          "update_follower": null,
          "election_is_running": null,
          "election_is_no_quorum": null,
          "election_is_finished": null,
          "update_election_is_running": null,
          "update_election_is_no_quorum": null,
          "update_election_is_finished": null,
          "initialized": null,
          "alive": true,
          "suspect": false,
          "confirmed": false,
          "persistent": true
        }
      },
      "member_id": "31efd0e3225b4e4999dee26ab5de3c9e"
    },
    "shield.default": {
      "population": {
        "3ba0a2e9c6cc40e5af8d444a419c635a": {
          "member_id": "3ba0a2e9c6cc40e5af8d444a419c635a",
          "service": "shield",
          "group": "default",
          "org": null,
          "cfg": {},
          "sys": {
            "ip": "172.26.0.3",
            "hostname": "9a84b2d54f9a",
            "gossip_ip": "172.26.0.3",
            "gossip_port": "9638",
            "http_gateway_ip": "0.0.0.0",
            "http_gateway_port": "9631"
          },
          "pkg": {
            "origin": "starkandwayne",
            "name": "shield",
            "version": "0.10.4",
            "release": "20170329112716"
          },
          "leader": null,
          "follower": null,
          "update_leader": null,
          "update_follower": null,
          "election_is_running": null,
          "election_is_no_quorum": null,
          "election_is_finished": null,
          "update_election_is_running": null,
          "update_election_is_no_quorum": null,
          "update_election_is_finished": null,
          "initialized": null,
          "alive": true,
          "suspect": false,
          "confirmed": false,
          "persistent": true
        }
      },
      "member_id": "31efd0e3225b4e4999dee26ab5de3c9e"
    }
  }
}
```

Shield endpoint:
```json
$ curl localhost:9632/census | jq
{
  "censuses": {
    "shield.default": {
      "population": {
        "3ba0a2e9c6cc40e5af8d444a419c635a": {
          "member_id": "3ba0a2e9c6cc40e5af8d444a419c635a",
          "service": "shield",
          "group": "default",
          "org": null,
          "cfg": {},
          "sys": {
            "ip": "172.26.0.3",
            "hostname": "9a84b2d54f9a",
            "gossip_ip": "0.0.0.0",
            "gossip_port": "9638",
            "http_gateway_ip": "0.0.0.0",
            "http_gateway_port": "9631"
          },
          "pkg": {
            "origin": "starkandwayne",
            "name": "shield",
            "version": "0.10.4",
            "release": "20170329112716"
          },
          "leader": null,
          "follower": null,
          "update_leader": null,
          "update_follower": null,
          "election_is_running": null,
          "election_is_no_quorum": null,
          "election_is_finished": null,
          "update_election_is_running": null,
          "update_election_is_no_quorum": null,
          "update_election_is_finished": null,
          "initialized": null,
          "alive": null,
          "suspect": null,
          "confirmed": null,
          "persistent": null
        }
      },
      "member_id": "3ba0a2e9c6cc40e5af8d444a419c635a"
    }
  }
}
```

**Comment by bodymindarts** [comment]
@eeyun was able to reproduce the behaviour but we have not yet found a minimal example to recreate it, nor are we sure of the root cause.

**Comment by reset** [comment]
@bodymindarts was able to give me a reproducable case which points to us skipping rebuilding the census when butterfly changes under particular circumstances. I should be able to get this one fixed up pretty soon unless I uncover some super nasty 👿

**Comment by reset** [comment]
Ok, I was able to reproduce the bug in a debugger and I tracked down the behaviour for why the rumors were being sent appropriate but the census was not being updated!

The issue is the `sleep` call on line 7 of the `init` hook in the shield plan. When we sleep in a hook we never return execution back to the supervisor program which means that we never have the ability to rebuild our census since that happens in the same thread of execution which the hook is being executed in.

I think the quick change is to recommend that people not sleep in these scripts at all and instead return a non zero exit code to signal back to the supervisor that the init script failed but it was normal and should be tried again. This will return execution flow back to the supervisor and allow the census to rebuilt getting us closer to a passing init hook.

I'm making the supervisor more concurrent in #1973 but even with that work we still can't have these hooks sleep - with that refactor the census would have been rebuilt, but the service thread still would never have be given back execution which means the config files wouldn't have been written out
